### PR TITLE
Filter logs based on revelancy

### PR DESCRIPTION
**Pull request by joeldudley**
### Why is this change needed?

Our log visibility model depends in part on the enclave only returning logs for which it has authenticated subscriptions.

### What changes were made as part of this PR:

- Implements `isRelevant`
- Reworks integration test in simulation to pass

### :rotating_light: Definition of done :rotating_light:
- [ ] Unit tests added to cover new or changed functionality 
- [ ] Docs pages updated to cover new or changed functionality
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated

### CODE REVIEW

File: go/enclave/rollupchain/rollup_chain.go [hunk 1]
```diff
@@ -241,7 +241,8 @@ func (rc *RollupChain) updateState(b *types.Block) (*obscurocore.BlockState, map
 	for _, receipt := range receipts {
 		logs = append(logs, receipt.Logs...)
 	}
-	subscribedLogs := rc.subscriptionManager.FilterRelevantLogs(logs)
+	// todo - joel - the creation of the state DB doesn't seem right
+	subscribedLogs := rc.subscriptionManager.FilterRelevantLogs(logs, rc.storage.CreateStateDB(head.Header.ParentHash))
```
Comment (joeldudley) [inline go/enclave/rollupchain/rollup_chain.go:245]: This is fine per Tudor, just need some special handling for the genesis rollup.

File: integration/simulation/simulation.go [hunk 2]
```diff
@@ -42,8 +42,8 @@ type Simulation struct {
 	SimulationTime   time.Duration
 	Stats            *stats.Stats
 	Params           *params.SimParams
-	LogChannel       chan types.Log
-	Subscriptions    []ethereum.Subscription
+	LogChannels      map[string]chan types.Log // Maps an owner to the channel on which they receive logs for all their wallets.
```
Comment (joeldudley) [inline integration/simulation/simulation.go:45]: Maintaining an owner -> channel mapping will be necessary once we start checking whether the node only receives relevant logs (this isn't possible currently, as we are missing the filtering by sub ID on the host side).

File: integration/simulation/validate_chain.go [hunk 3]
```diff
@@ -378,11 +377,19 @@ func checkLogsReceived(t *testing.T, s *Simulation) {
 		sub.Unsubscribe()
 	}
 
+	for _, channel := range s.LogChannels {
+		checkReceivedHOCAndPOCLogs(t, channel)
```
Comment (joeldudley) [inline integration/simulation/validate_chain.go:381]: We now check that each account has received HOC and POC events, and not the overall network.

File: go/enclave/events/subscription_manager.go [hunk 4]
```diff
@@ -97,12 +100,25 @@ func (s *SubscriptionManager) EncryptLogs(logsBySubID map[uuid.UUID][]*types.Log
...
-func isRelevant(log *types.Log, sub *common.LogSubscription) (bool, *common.SubscriptionAccount) {
-	// Work out if this is an account or lifecycle log
-	// If the former, establish whether it is relevant to any subscription
-	// Return the first account for which the log matches, so it can be used for encryption
-	return true, nil
+// Indicates whether the log is relevant for the subscription. A lifecycle log is considered relevant to everyone.
+func isRelevant(log *types.Log, sub *common.LogSubscription, db *state.StateDB) bool {
+	var nonContractAddrs []string
+	for _, topic := range log.Topics {
+		addr := gethcommon.HexToAddress(topic.Hex())
```
Comment (tudor-malene) [inline go/enclave/events/subscription_manager.go:107]: it would make sense to filter out topics that are not exactly 20 bytes

**Review by tudor-malene — COMMENTED** [review]
looks good in principle
few comments

File: go/enclave/events/subscription_manager.go [hunk 5]
```diff
@@ -97,12 +100,25 @@ func (s *SubscriptionManager) EncryptLogs(logsBySubID map[uuid.UUID][]*types.Log
...
+		if db.GetCode(addr) == nil { // If there is code associated with the address, it's a contract address.
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/enclave/events/subscription_manager.go:108]: Fixed

File: go/enclave/events/subscription_manager.go [hunk 6]
```diff
@@ -97,12 +100,25 @@ func (s *SubscriptionManager) EncryptLogs(logsBySubID map[uuid.UUID][]*types.Log
...
+			nonContractAddrs = append(nonContractAddrs, addr.Hex())
+		}
+	}
+
+	if len(nonContractAddrs) == 0 {
+		// All the topic addresses are contract addresses, so this is a log event, and is relevant to everyone.
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/enclave/events/subscription_manager.go:114]: Fixed

File: go/enclave/events/subscription_manager.go [hunk 7]
```diff
@@ -97,12 +100,25 @@ func (s *SubscriptionManager) EncryptLogs(logsBySubID map[uuid.UUID][]*types.Log
...
+		return true
+	}
+
+	for _, addr := range nonContractAddrs {
+		if addr == sub.SubscriptionAccount.Account.Hex() {
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/enclave/events/subscription_manager.go:119]: Fixed

File: go/enclave/events/subscription_manager.go [hunk 8]
```diff
@@ -97,12 +100,25 @@ func (s *SubscriptionManager) EncryptLogs(logsBySubID map[uuid.UUID][]*types.Log
...
 
-// Indicates whether the log is relevant for any subscription, and returns the matching subscription account if so.
-// A lifecycle log is considered relevant to everyone.
-// TODO - #453 - Filter logs, instead of considering all logs relevant to everyone.
-func isRelevant(log *types.Log, sub *common.LogSubscription) (bool, *common.SubscriptionAccount) {
-	// Work out if this is an account or lifecycle log
-	// If the former, establish whether it is relevant to any subscription
-	// Return the first account for which the log matches, so it can be used for encryption
-	return true, nil
+// Indicates whether the log is relevant for the subscription. A lifecycle log is considered relevant to everyone.
```
[the remark this answers is not in the document]
Comment (joeldudley) [reply go/enclave/events/subscription_manager.go:103]: Fixed

File: go/enclave/events/subscription_manager.go (hunk 4 above)
Comment (joeldudley) [reply go/enclave/events/subscription_manager.go:107]: Good idea, thanks

**Review by tudor-malene — APPROVED** [review]
lgtm

**Pull request merged**